Custom Host Mac/Linux

aishwaryabh opened the pull request:
<!-- Please provide all the information below.  -->

### Issue describing the changes in this PR

resolves [#3886 ](https://github.com/Azure/azure-functions-core-tools/issues/3886)

This PR creates a custom host for mac and linux. This PR does not cover signing the artifacts and simply generates them and modifies the artifact assembler.

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)

kshyju commented on eng/ci/templates/official/jobs/build-core-tools-host.yml:
```diff
@@ -35,89 +35,138 @@ jobs:
+    displayName: Install .NET 9
+  - task: UseDotNet@2
+    inputs:
+      version: 6.x
```
Do we need 6.x ?

aishwaryabh replied:
yeah the signing task needs .NET 6

kshyju commented on eng/ci/templates/official/jobs/build-core-tools-host.yml:
```diff
@@ -35,89 +35,138 @@ jobs:
+      command: publish
+      publishWebProjects: false
+      zipAfterPublish: false
+      arguments: -c Release -r osx-x64 -o $(Build.SourcesDirectory)/pkg_output/osx/osx-x64 --self-contained
```
Do we need this `--self-contained` argument?

aishwaryabh replied:
I think so we needed it for windows so prob need for linux/mac?

kshyju commented on eng/ci/templates/official/jobs/build-core-tools-host.yml:
```diff
@@ -35,89 +35,138 @@ jobs:
-      artifact: drop-coretools-host-windows  
\ No newline at end of file
+      path: $(Build.ArtifactStagingDirectory)/_coreToolsHostPackages
+      artifact: drop-coretools-host-mac
```
nit: should we use "osx" instead of "mac" ?

kshyju commented on src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs:
```diff
@@ -14,7 +14,10 @@ internal sealed class ArtifactAssembler
         private readonly Dictionary<string, string> _customHostArtifacts = new()
```
Should we consider renaming this variable to "_visualStudioArtifacts"?

kshyju commented:
Can you rebase this PR branch?

aishwaryabh commented:
This PR was WAY too messy with the force push of `feature/oop-host` to rebase. I've decided to close this PR and open this one instead. I have addressed the comments there as well.

aishwaryabh commented on src/Azure.Functions.ArtifactAssembler/EnvironmentVariables.cs:
```diff
@@ -10,7 +10,9 @@ internal static class EnvironmentVariables
-        internal const string CoreToolsHostArtifactName = "CORETOOLS_HOST_ARTIFACT_NAME";
+        internal const string CoreToolsHostWindowsArtifactName = "CORETOOLS_HOST_WINDOWS_ARTIFACT_NAME";
+        internal const string CoreToolsHostLinuxArtifactName = "CORETOOLS_HOST_LINUX_ARTIFACT_NAME";
+        internal const string CoreToolsHostMacArtifactName = "CORETOOLS_HOST_MAC_ARTIFACT_NAME";
```
already done!
